### PR TITLE
feat: Add semver range support to all cmd & pkg

### PR DESCRIPTION
**Pull request by viccuad**
Use `"github.com/Masterminds/semver/v3"` to validate, honor, and use the semver constraints listed in shared deps everywhere.
Which right now means:
- `hypper lint CHART`: semver constraint is now a must (we will ERROR if not present).
- `hyper shared-dep list CHART`:  filter against the constraint correctly, now outputting `status: out-of-range` if the installed dep doesn't satisfy what is needed.
- `hypper install CHART`: pull correct chart that satisfies the needed semver constraint. Error if the repository/cache doesn't contain a version that can satisfy it.

In addition, 
- Add related tests. For that, add new `pkg/chart/{Mock(),MockChartOptions()}` functions, since tests for dependencies need releases with charts that contain specific `Chart.Metadata.Version`. I suppose these functions will grow in the future, and we should think about upstreaming them.
- Fix a bug with ns when `hypper shared-dep list -n foo`.

Closes: https://github.com/rancher-sandbox/hypper/issues/90

### CODE REVIEW

**Comment by viccuad** [comment]
I haven't updated docs/ with the slightly different ouputs.
(also, it may be a good idea to review the PR stepping through commits, instead of by files).

File: cmd/hypper/testdata/output/install-with-shared-deps-out-of-range.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+🛳  Installing shared dependencies for chart "empty":
+ERROR: ❌  Satisfiable version for chart "testdata/testcharts/shared-dep" not found, aborting
+Error: Satisfiable chart version not found; 0.1.0 is less than 0.3.0
```
Comment (viccuad) [inline cmd/hypper/testdata/output/install-with-shared-deps-out-of-range.txt:3]: I don't like this output at all;
I thought that line 2 should just be red (it is) but not contain the `ERROR` in the beginning.
Also, line 3 is a legit error that has been floated upwards, I suppose we are forgetting to print it through our logger. I suppose we can fix it in a following PR.

File: cmd/hypper/install.go [hunk 2]
```diff
@@ -65,7 +65,7 @@ func newInstallCmd(actionConfig *action.Configuration, logger log.Logger) *cobra
 			if err != nil {
 				return err
 			}
-			logger.Info(eyecandy.ESPrint(settings.NoEmojis, ":clapping_hands: Done!"))
+			logger.Info(eyecandy.ESPrint(settings.NoEmojis, ":clapping_hands:Done!"))
```
Comment (dragonchaser) [inline cmd/hypper/install.go:68]: Why did you remove the space? Looks better with....

File: cmd/hypper/testdata/output/install-with-shared-deps-out-of-range.txt (hunk 1 above)
Comment (dragonchaser) [reply cmd/hypper/testdata/output/install-with-shared-deps-out-of-range.txt:3]: You are right, we should add a cart for that (e.g. expand the style guide to include colouring of output).

**Review by dragonchaser — APPROVED** [review]

File: cmd/hypper/install.go (hunk 2 above)
Comment (viccuad) [reply cmd/hypper/install.go:68]: it's unclear from the code, but with this change, it looks aligned in my terminal:

tty

Curiously, it looks also aligned on emacs:

emacs

But it doesn't look aligned in vscode:

vscode

nor if I copy-paste the testdata/output and format it here:

```
🛳  Installing shared dependencies for chart "empty":
🛳  - Installing chart "shared-dep-empty" as "my-shared-dep" in namespace "my-shared-dep-ns"…
🛳  Installing chart "empty" as "my-hypper-name" in namespace "hypper"…
👏 Done!
```

I think the clapping-hands emoji is 2 characters wide, and in some places it gets shown in one way, and in others, the other way. What a mess! :man_shrugging: 

I wonder how the ouput from this PR looks in other terminals besides mine! here in xfce and gnome term, it looks ok with the PR change xD.

Comment (dragonchaser) [reply cmd/hypper/install.go:68]: :shrug: not a blogger for me, just wanted to point that out, I guess output largely depends on the used terminal....

File: cmd/hypper/testdata/output/install-with-shared-deps-out-of-range.txt (hunk 1 above)
Comment (viccuad) [reply cmd/hypper/testdata/output/install-with-shared-deps-out-of-range.txt:3]: opened https://github.com/rancher-sandbox/hypper/issues/155 !

**Comment by viccuad** [comment]
merging! thanks :)
